zerver/lib/stream_subscription.py

```diff
@@ -246,3 +246,33 @@ def handle_stream_notifications_compatibility(
         stream_dict[notification_type] = (
             False if user_profile is None else getattr(user_profile, target_attr)
         )
+
+
+def user_ids_with_stream_history_access(stream: Stream) -> Set[int]:
+    user_ids: Set[int] = set()
+    # The history access has 3 conditions:
+    # 1. if !history_public_to_subscribers:
+    #       History is not available to anyone
+    # 2. if history_public_to_subscribers and is_web_public:
+    #       All subscribers can access the history including guests
+    # 3. if history_public_to_subscribers and !is_web_public:
+    #       All subscribers can access the history excluding guests
+
+    if not stream.is_history_public_to_subscribers():
+        return user_ids
+
+    subscriptions = get_active_subscriptions_for_stream_id(stream.id)
+    if stream.is_history_public_to_subscribers() and stream.is_web_public:
+        user_ids = set(
+            subscriptions.filter(user_profile__is_active=True).values_list(
+                "user_profile__id", flat=True
+            )
+        )
+    elif stream.is_history_public_to_subscribers() and not stream.is_web_public:
+        user_ids = set(
+            subscriptions.filter(user_profile__is_active=True)
+            .exclude(user_profile__role=600)
+            .values_list("user_profile__id", flat=True)
+        )
```

Comment:
This is a confusing way to write a function; there's no point in the `if/else` statements, especially rechecking `is_history_public_to_subscribers` in each one, if we return after the first one.

I think you just want to remove the `user_ids` variable and just use `return` statements in each case.

Reply:
Ok, will update the PR soon. I thought of writing this way so that anyone trying to understand the code finds it easier to do so.

Reply:
Fine, you seemed to merged this commit, Thanks.